src/deploy/functions/params.ts

```diff
@@ -0,0 +1,342 @@
+  param: string;
+
+  // A human friendly name for the param. Will be used in install/configure flows to describe
+  // what param is being updated. If omitted, UX will use the value of "param" instead.
+  label?: string;
+
+  // A long description of the parameter's purpose and allowed values. If omitted, UX will not
+  // provide a description of the parameter
+  description?: string;
+
+  // Default value. If not provided, a param must be supplied.
+  default?: T | build.Expression<T>;
+
+  // default: false
+  immutable?: boolean;
+}
+
+export interface StringParam extends ParamBase<string> {
+  type: "string";
+
+  // If omitted, defaults to TextInput<string>
+  input?: TextInput<string> | SelectInput<string>;
+}
+
+export interface IntParam extends ParamBase<number> {
+  type: "int";
+
+  // If omitted, defaults to TextInput<number>
+  input?: TextInput<number> | SelectInput<number>;
+}
+
+export interface TextInput<T, Extensions = {}> { // eslint-disable-line
+  type?: "text";
+
+  text:
+    | Extensions
+    | {
+        example?: string;
+      };
+}
+
+interface SelectOptions<T> {
+  // Optional human-facing value for this option (e.g. "US Central (Iowa)" instead of value
+  // "us-central1")
+  label?: string;
+
+  // Actual value of the parameter if this option is selected
+  value: T;
+}
+
+export interface SelectInput<T> {
+  type?: "select";
+
+  select: Array<SelectOptions<T>>;
+}
+
+export type Param = StringParam | IntParam;
+type ParamValue = string | number | boolean;
+
+/**
+ * Calls the corresponding resolveX function for the type of a param.
+ * To be used when resolving the default value of a param, if CEL.
+ * It's an error to call this on a CEL expression that depends on params not already known in the currentEnv.
+ */
+function resolveDefaultCEL(
+  type: string,
+  expr: CEL,
+  currentEnv: Record<string, ParamValue>
+): ParamValue {
+  const deps = dependenciesCEL(expr);
+  const allDepsFound = deps.every((dep) => {
+    return Object.keys(currentEnv).includes(dep);
+  });
+  if (!allDepsFound) {
+    throw new FirebaseError(
+      "Build specified parameter with un-resolvable default value " +
+        expr +
+        "; dependencies missing."
+    );
+  }
+
+  switch (type) {
+    case "string":
+      return resolveString(expr, currentEnv);
+    case "int":
+      return resolveInt(expr, currentEnv);
+    default:
+      throw new FirebaseError(
+        "Build specified parameter with default " + expr + " of unsupported type"
+      );
+  }
+}
+/**
+ * Tests whether a mooted ParamValue literal is of the correct type to be the value for a Param.
+ */
+function canSatisfyParam(param: Param, value: ParamValue): boolean {
+  if (param.type === "string") {
+    return typeof value === "string";
+  } else if (param.type === "int") {
+    return typeof value === "number" && Number.isInteger(value);
+  } else {
+    assertExhaustive(param);
```

Comment:
nit: you don't need to put this in an else. The only time you'd put it in a block would be if you wanted to do a select statement:

```
select(param.type) {
case "string": return typeof value === "string";
case "int": return typeof value === "int" && Number.isInteger(value);
default: assertExhaustive();
}
```

Even then, I don't know if you need to put it in a default. We're using a special `never` type to hint to the compiler that this piece of code will never run. It helps the compiler recognize that enum types have been all matched and therefore we've seen all code paths that can happen.

Reply:
Ah, so it's basically just checking if the compiler thinks that the type of the parameter is `never`?